static_frame/core/hloc.py

```diff
@@ -22,11 +24,53 @@ class HLoc(metaclass=HLocMeta):
             'key',
             )
 
-    def __init__(self, key: tp.Sequence[GetItemKeyType]):
+    def __init__(self, key: tp.Tuple[GetItemKeyType]) -> None:
         self.key = key
 
     def __iter__(self) -> tp.Iterator[GetItemKeyType]:
         return self.key.__iter__()
 
     def __len__(self) -> int:
         return self.key.__len__()
+
+    def __repr__(self) -> str:
+        contains_slices = False
+
+        # This is usually very small (i.e. <10 items), so it's very cheap to
+        # pay the price of an extra loop to check for slices.
+        for key in self.key:
+            if key.__class__ is slice:
+                contains_slices = True
+                break
+
+        def gen_nested_keys() -> tp.Iterator[str]:
+            for key in self.key:
+                if not isinstance(key, slice):
+                    yield str(key)
+                    continue
+
+                if key == NULL_SLICE:
+                    yield ':'
+                    continue
+
+                if key.start is None:
+                    result = ':'
+                else:
+                    result = f'{key.start}:'
+
+                if key.stop is not None:
+                    result += str(key.stop)
+
+                if key.step is not None and key.step != 1:
+                    result += f':{key.step}'
+
+                yield result
+
+        if not contains_slices:
+            if len(self.key) == 1:
+                return f'HLoc[{self.key[0]}]'
+
+            # self.key is a tuple, so we strip off the parentheses.
+            return f'HLoc[{str(self.key)[1:-1]}]'
+
+        return f'HLoc[{",".join(gen_nested_keys())}]'
```

Comment:
Should we do something like this?

```suggestion
        return f'<HLoc[{",".join(gen_nested_keys())}]>'
```

Reply:
I'm not sure I like this actually. It feels like it makes the repr unnecessarily artificial. Can you help me understand why you want this?

Reply:
I do not have a strong opinion, but it is common to put class names in brackets, e.g.:

```
>>> sf.Series((1,2))                                                                                                                                                                                                                                                           
<Series>
<Index>
0        1
1        2
<int64>  <int64>
```
Unless you expect your repr to actually be "evalable"... but we do not have a lot of examples of  reprs for non-containers in StaticFrame.